File: Algorithms/C++/Medium/364_Nested_List_Weight_Sum_II.cpp
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */
class Solution {
public:
    int depthSumInverse(vector<NestedInteger>& nestedList) {
        int deepestLevel = getDeepestLevel(nestedList, 1);

        return getDepthSum(nestedList, deepestLevel);
    }
private:
    int getDeepestLevel(vector<NestedInteger>& nestedList, int currentLevel) {
        int deepestLevel = currentLevel;

        for (int i = 0; i < nestedList.size(); i++) {
            if (!nestedList[i].isInteger()) {
                int level = getDeepestLevel(nestedList[i].getList(), currentLevel + 1);
                if (level > deepestLevel) {
                    deepestLevel = level;
                }
            }
        }

        return deepestLevel;
    }

    int getDepthSum(vector<NestedInteger>& nestedList, int currentLevel) {
        int sum = 0;
        for (int i = 0; i < nestedList.size(); i++) {
            if (!nestedList[i].isInteger()) {
                sum += getDepthSum(nestedList[i].getList(), currentLevel - 1);
            } else {
                sum += nestedList[i].getInteger() * currentLevel;
            }
        }

        return sum;
    }
};

File: Algorithms/Swift/Medium/364_Nested_List_Weight_Sum_II.swift
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public func isInteger() -> Bool
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     public func getInteger() -> Int
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     public func getList() -> [NestedInteger]
 * }
 */
class Solution {
    func depthSumInverse(nestedList: [NestedInteger]) -> Int {
        if nestedList.isEmpty {
            return 0
        }

        let deepestLevel = getDeepestLevelFor(nestedList, withCurrentLevel: 1)

        return getDepthSumFor(nestedList, withCurrentLevel: deepestLevel)
    }

    private func getDeepestLevelFor(nestedList: [NestedInteger], withCurrentLevel currentLevel: Int) -> Int {
        var deepestLevel = currentLevel

        for item in nestedList {
            if !item.isInteger() {
                let level = getDeepestLevelFor(item.getList(), withCurrentLevel: currentLevel + 1)
                if level > deepestLevel {
                    deepestLevel = level
                }
            }
        }

        return deepestLevel
    }

    private func getDepthSumFor(nestedList: [NestedInteger], withCurrentLevel currentLevel: Int) -> Int {
        var sum = 0

        for item in nestedList {
            if item.isInteger() {
                sum += item.getInteger() * currentLevel
            } else {
                sum += getDepthSumFor(item.getList(), withCurrentLevel: currentLevel - 1)
            }
        }

        return sum
    }
}
